src/sources/socket.rs

```diff
@@ -0,0 +1,233 @@
+use super::util::TcpSource;
+use crate::{
+    event::{self, Event},
+    sources::tcp::{RawTcpSource, TcpConfig},
+    sources::udp::{udp, UdpConfig},
+    topology::config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+};
+use futures::sync::mpsc;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+// TODO: add back when serde-rs/serde#1358 is addressed
+// #[serde(deny_unknown_fields)]
+pub struct SocketConfig {
+    #[serde(flatten)]
+    pub mode: Mode,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(tag = "mode", rename_all = "snake_case")]
+pub enum Mode {
+    Tcp(TcpConfig),
+    Udp(UdpConfig),
+}
+
+impl SocketConfig {
+    pub fn new(mode: Mode) -> Self {
+        Self { mode }
+    }
+}
+
+inventory::submit! {
+    SourceDescription::new_without_default::<SocketConfig>("socket")
+}
+
+#[typetag::serde(name = "socket")]
+impl SourceConfig for SocketConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        match self.mode.clone() {
+            Mode::Tcp(config) => {
+                let tcp = RawTcpSource {
+                    config: config.clone(),
+                };
+                tcp.run(config.address, config.shutdown_timeout_secs, out)
+            }
+            Mode::Udp(config) => {
+                let host_key = config.host_key.clone().unwrap_or(event::HOST.clone());
+                Ok(udp(config.address, host_key, out))
+            }
+        }
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "socket"
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
These tests are all just the existing tcp/udp source tests moved into this file.  There's no new test functionality.  The only new test I could imagine writing would be something about the config file parsing behavior to make sure it requires the 'mode' field, errors on unexpected values, etc, but I didn't see any other tests testing the parser like that.  If you think this is worth doing just lmk.

Reply:
We have had issues in the past with serde annotations not doing what was expected. As such, testing parsing behavior does make sense, though I would expect serde to be fairly robust itself.

Reply:
So after I made this comment I did actually find some explicit parser tests in tests/config.rs and added a test that the 'mode' field is now required for sources of type 'socket'.  Please let me know if you think that is sufficient or if there's something else you think is worth testing.